### ISSUE TITLE
Design meeting notes - July 23, 2015

### ISSUE THREAD

**Issue by ajcvickers** [dotnet/efcore#2706]
# NuGet/DNX Commands

We reviewed the Entity Framework Commands and their parameters. Here is the outcome.

NuGet | DNX
----- | ---
Add-Migration<br />&nbsp;&nbsp;&nbsp;&nbsp;[Name]<br />&nbsp;&nbsp;&nbsp;&nbsp;-Context<br />&nbsp;&nbsp;&nbsp;&nbsp;-Project<br />&nbsp;&nbsp;&nbsp;&nbsp;-StartupProject | dnx [project] ef migration**s** add<br />&nbsp;&nbsp;&nbsp;&nbsp;[name]<br />&nbsp;&nbsp;&nbsp;&nbsp;-c, --context<br />&nbsp;&nbsp;&nbsp;&nbsp;-s, --startupProject
Remove-Migration<br />&nbsp;&nbsp;&nbsp;&nbsp;-Context<br />&nbsp;&nbsp;&nbsp;&nbsp;-Project<br />&nbsp;&nbsp;&nbsp;&nbsp;-StartupProject | dnx [project] ef migration**s** remove<br />&nbsp;&nbsp;&nbsp;&nbsp;-c, --context<br />&nbsp;&nbsp;&nbsp;&nbsp;-s, --startupProject
Script-Migration<br />&nbsp;&nbsp;&nbsp;&nbsp;-From *(required when -To)*<br />&nbsp;&nbsp;&nbsp;&nbsp;-To<br />&nbsp;&nbsp;&nbsp;&nbsp;-Idempotent<br />&nbsp;&nbsp;&nbsp;&nbsp;-Context<br />&nbsp;&nbsp;&nbsp;&nbsp;-Project<br />&nbsp;&nbsp;&nbsp;&nbsp;-StartupProject | dnx [project] ef migration**s** script<br />&nbsp;&nbsp;&nbsp;&nbsp;~~[from]~~ **-f, --from** *(required when --to)*<br />&nbsp;&nbsp;&nbsp;&nbsp;~~[to]~~ **-t, --to**<br />&nbsp;&nbsp;&nbsp;&nbsp;-i, --idempotent<br />&nbsp;&nbsp;&nbsp;&nbsp;-o, --output<br />&nbsp;&nbsp;&nbsp;&nbsp;-c, --context<br />&nbsp;&nbsp;&nbsp;&nbsp;-s, --startupProject
~~Apply-Migration~~ **Update-Database**<br />&nbsp;&nbsp;&nbsp;&nbsp;[Migration]<br />&nbsp;&nbsp;&nbsp;&nbsp;-Context<br />&nbsp;&nbsp;&nbsp;&nbsp;-Project<br />&nbsp;&nbsp;&nbsp;&nbsp;-StartupProject | dnx [project] ef ~~migration apply~~ **database update**<br />&nbsp;&nbsp;&nbsp;&nbsp;[migration]<br />&nbsp;&nbsp;&nbsp;&nbsp;-c, --context<br />&nbsp;&nbsp;&nbsp;&nbsp;-s, --startupProject
~~Reverse-Engineer~~ **Scaffold-DbContext**<br />&nbsp;&nbsp;&nbsp;&nbsp;[Connection~~String~~]<br />&nbsp;&nbsp;&nbsp;&nbsp;[Provider]<br />&nbsp;&nbsp;&nbsp;&nbsp;-Project<br />&nbsp;&nbsp;&nbsp;&nbsp;~~-StartupProject~~ | dnx [project] ef ~~revEng~~ **dbcontext scaffold**<br />&nbsp;&nbsp;&nbsp;&nbsp;[connection~~String~~]<br />&nbsp;&nbsp;&nbsp;&nbsp;[provider]
~~Customize-ReverseEngineer~~ **Scaffold-DbContextTemplate**<br />&nbsp;&nbsp;&nbsp;&nbsp;[Provider]<br />&nbsp;&nbsp;&nbsp;&nbsp;-Project<br />&nbsp;&nbsp;&nbsp;&nbsp;~~-StartupProject~~ | dnx [project] ef ~~revEng customize~~ **dbcontext scaffold-templates**<br />&nbsp;&nbsp;&nbsp;&nbsp;[provider]
Use-DbContext<br />&nbsp;&nbsp;&nbsp;&nbsp;[Context]<br />&nbsp;&nbsp;&nbsp;&nbsp;-Project | *(no session)*
*(tab expansion)* | dnx [project] ef **db**context list<br />&nbsp;&nbsp;&nbsp;&nbsp;-s, --startupProject
*(tab expansion)* | dnx [project] ef migration**s** list<br />&nbsp;&nbsp;&nbsp;&nbsp;-s, --startupProject

Here are some other, general notes we had.
* Cross-reference related commands in help
* For renamed commands, ensure errors are useful to early adopters

It was also interesting to view the DNX sub-commands as a hierarchy.

* migrations
    * list
    * add
    * remove
    * script
* dbcontext
    * list
    * scaffold
    * scaffold-templates
* database
    * update

# Cross-platform coding guidelines
We reviewed and discussed how to make the EF code more sensitive to Linux and OS X. The issues discussed have now been addressed in [Engineering Guidelines - Cross-platform coding](https://github.com/aspnet/Home/wiki/Engineering-guidelines#cross-platform-coding).

# Discussion

Please comment on this issue to provide feedback, ask questions, etc.

**Comment by bricelam** [comment]
We finally had to let go of our revEng...

**Comment by Muchiachio** [comment]
Is -force switch missing or it's no longer needed in EF7?

**Comment by bricelam** [comment]
On `Update-Database`? It is no longer needed due to [the removal of automatic migrations](http://www.bricelam.net/2014/12/16/ef7-migrations-designtime.html).

**Comment by Muchiachio** [comment]
Yes on `Update-Database`, I was wondering because calling `Update-Database -TargetMigration PreviousMigration` would most of the time result in data loss and it was stopped by an error in EF6.

**Comment by bricelam** [comment]
Oh weird; I thought we only blocked data loss on automatic migrations. In EF7, we'll warn you that they have been generated in the `Up` method during `Add-Migration`.

**Comment by Muchiachio** [comment]
Here are some other things I noticed :sweat_smile:
```
dnx [project] ef migrations script
    [from] -f, --from (required when --to)
    [to] -t, --to
    -i, --idempotent
    -c, --context
    -s, --startupProject
```
Definitely liking the **s**'es, seems to be more natural to read. :+1:
But why is --from required when --to is specified? I could think of two scenarios when only one of these parameters would be specified.
- I am developing a system, and I just deploy as is, meaning I only need not applied migrations. I go to command line and run `dnx . ef migrations script --from FirstMigrationFor_v1.1`. This works fine now.
- I am developing a system, it already has a tagged v1.0 and v1.1-beta releases. I passed all the QA madness and I'm ready to deploy v1.0 on production. I need scripts so I go to command line and run `dnx . ef migrations script --to LastMigrationFor_v1` (--upto switch would sound even better). This would mean *"Generate all the script needed for v1.0 release"*.

I read your [blog post](http://www.bricelam.net/2014/12/16/ef7-migrations-designtime.html) and I couldn't help but notice separating migrations by namespace/folder. Thus having a -n, --namespace switch would be nice too. This could be used to generate version specific scripts like `dnx . ef migrations script --namespace MyProject.Migration.V1_1`. It could be a nice wrapper for ranging --from --to commands, but it's not necessary for initial release :smiley:.

```
dnx [project] ef migrations list
    -s, --startupProject
```
It would be nice to have a "not applied" option later on.
And as with migrations --namespace switch to list migrations in a given namespace.
